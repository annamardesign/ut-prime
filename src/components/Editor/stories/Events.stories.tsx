import React from 'react';
import type { Story } from '@storybook/react';
import type { Props } from '../Editor.types';
import Editor from '..';
import useToast from '../../hooks/useToast';

import {input, dropdowns} from '../../test/input';

export {default} from '../Editor.stories';

const sticky = {sticky: false};

const Template: Story<Props> = ({methods, ...args}) => {
    const {toast, submit} = useToast(sticky);
    return <>
        <Editor
            id={1}
            object='input'
            onGet={() => Promise.resolve({input: {}})}
            onDropdown={() => Promise.resolve(dropdowns)}
            onAdd={submit}
            onEdit={submit}
            onFieldChange='handleFieldChange'
            methods={{
                async handleFieldChange({field, value, event}: {field: unknown, value, event: Event}) {
                    submit({field, value});
                    // throw new Error('test error');
                    // return false;
                },
                async handleAutocomplete() {
                    return [{
                        value: 1, label: 'value 1'
                    }, {
                        value: 2, label: 'value 2'
                    }, {
                        value: 3, label: 'value 3'
                    }];
                },
                async 'portal.customization.get'() {
                    return {};
                },
                ...methods
            }}
            {...args}
        />
        {toast}
    </>;
};

declare type StoryTemplate = Story<Partial<Props>> & {
    play: (context: {canvasElement: HTMLElement}) => Promise<void>
}

export const Events: StoryTemplate = Template.bind({});
Events.args = {
    ...input,
    layouts: {edit: ['left', 'center', 'right']}
};

export const EventsFormAPI: StoryTemplate = Template.bind({});
EventsFormAPI.args = {
    schema: {
        properties: {
            a: {
                type: 'integer',
                widget: {onChange: 'handleA'}
            },
            b: {
                type: 'integer',
                widget: {onChange: 'handleB'}
            },
            sum: {
                type: 'number',
                readOnly: true
            },
            discount: {
                type: 'number',
                widget: {onChange: 'handleDiscount'}
            },
            bitmask: {
                widget: {onChange: 'handleBitmask'}
            }
        }
    },
    cards: {
        edit: {
            className: 'lg:col-3',
            widgets: [
                'a',
                'b',
                'sum',
                'discount',
                'bitmask'
            ]
        }
    },
    onGet: () => Promise.resolve({a: 0, b: 0, sum: 0}),
    methods: {
        async handleA({form, value}) {
            form.setValue('sum', Number(form.getValues('b')) + Number(value));
        },
        async handleB({form, value}) {
            form.setValue('sum', Number(form.getValues('a')) + Number(value));
        },
        async handleDiscount({value, form}) {
            if (Number(value) < 0) throw new Error('Discount must be > 0'); // throwing means abort the edit and show error
            if (Number(value) >= 0) form.clearErrors('discount');
        },
        async handleBitmask({value}) {
            if (/[^01]/.test(value)) return false; // returning false means abort the edit silently
        }
    }
};
