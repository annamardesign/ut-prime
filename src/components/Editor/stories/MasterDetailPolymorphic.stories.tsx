import { userEvent, within } from '@storybook/testing-library';
import React from 'react';
import Editor from '..';
export {default} from '../Editor.stories';

const common = {
    className: 'xl:col-3',
    label: 'Shape properties',
    watch: '$.selected.shape'
};

export const MasterDetailPolymorphic = () =>
    <Editor
        id={1}
        onGet={() => Promise.resolve({
            shape: [
                {name: 'Big circle', type: 'circle', radius: 100},
                {name: 'Small ellipse', type: 'ellipse', radius: 5, secondRadius: 4},
                {name: 'Big square', type: 'square', side1: 100},
                {name: 'Medium triangle', type: 'triangle', side1: 30, side2: 40, side3: 50},
                {name: 'Small rectangle', type: 'rectangle', side1: 2, side2: 4}
            ]
        })}
        onDropdown={() => Promise.resolve({})}
        schema={{
            type: 'object',
            properties: {
                shape: {
                    type: 'array',
                    items: {
                        type: 'object',
                        properties: {
                            name: {},
                            type: {enum: ['circle', 'square', 'triangle', 'rectangle', 'ellipse']},
                            radius: {title: 'Radius', type: 'number'},
                            secondRadius: {title: 'Second radius', type: 'number'},
                            side1: {title: 'Side Length', type: 'number'},
                            side2: {title: 'Second Side Length', type: 'number'},
                            side3: {title: 'Third Side Length', type: 'number'},
                            tags: {}
                        }
                    },
                    title: '',
                    widget: {
                        type: 'table',
                        selectionMode: 'single',
                        actions: {
                            allowEdit: false
                        }
                    }
                }
            }
        }}
        cards={{
            shapes: {
                label: 'Shapes',
                className: 'xl:col-2',
                widgets: [{
                    name: 'shape',
                    widgets: ['name']
                }]
            },
            shapeCircle: {
                ...common,
                match: {type: 'circle'},
                widgets: [
                    '$.edit.shape.radius'
                ]
            },
            shapeEllipse: {
                ...common,
                match: {type: 'ellipse'},
                widgets: [
                    '$.edit.shape.radius',
                    '$.edit.shape.secondRadius'
                ]
            },
            shapeSquare: {
                ...common,
                match: {type: 'square'},
                widgets: [
                    '$.edit.shape.side1'
                ]
            },
            shapeRectangle: {
                ...common,
                match: {type: 'rectangle'},
                widgets: [
                    '$.edit.shape.side1',
                    '$.edit.shape.side2'
                ]
            },
            shapeTriangle: {
                ...common,
                match: {type: 'triangle'},
                widgets: [
                    '$.edit.shape.side1',
                    '$.edit.shape.side2',
                    '$.edit.shape.side3'
                ]
            }
        }}
        layouts={{
            edit: ['shapes', ['shapeCircle', 'shapeSquare', 'shapeTriangle', 'shapeRectangle', 'shapeEllipse']]
        }}
    />;

MasterDetailPolymorphic.play = async({canvasElement}) => {
    const canvas = within(canvasElement);
    userEvent.click(canvas.getByText('Small ellipse'));
};
