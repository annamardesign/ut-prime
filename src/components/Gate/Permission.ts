import PropTypes from 'prop-types';
import { Component } from 'react';

class Permission extends Component {
    static childContextTypes = {
        translate: PropTypes.func,
        money: PropTypes.func,
        dateFormat: PropTypes.func,
        numberFormat: PropTypes.func,
        checkPermission: PropTypes.func,
        implementationStyle: PropTypes.object
    };

    getChildContext() {
        return {
            translate: text => text,
            dateFormat: text => text,
            numberFormat: text => text,
            checkPermission: action => true,
            money: text => text,
            implementationStyle: {}
        };
    }

    render() {
        return this.props.children;
    }
}

export default Permission;
