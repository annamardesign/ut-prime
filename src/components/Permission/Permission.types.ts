import React from 'react';

export interface Props {
    permission: string | string[];
}

export type StyledType = React.FC<Props>
