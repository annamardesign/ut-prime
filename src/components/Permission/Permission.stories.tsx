import React from 'react';
import type { Story, Meta } from '@storybook/react';

// @ts-ignore: md file and not a module
import page from './README.mdx';
import Permission from './index';
import type { Props } from './Permission.types';

import state from '../test/state';

const meta: Meta = {
    title: 'Permission',
    component: Permission,
    parameters: {docs: {page}},
    args: {
        state
    }
};
export default meta;

const Template: Story<Props & {state: {}}> = ({state, ...args}) =>
    <Permission {...args}>
        <div className='p-component'>
            Permissions checked: {String(args.permission)}
        </div>
    </Permission>;

export const Basic: Story<Props> = Template.bind({});
Basic.args = {
    permission: undefined
};

export const Granted: Story<Props> = Template.bind({});
Granted.args = {
    ...Basic.args,
    permission: 'granted'
};

export const Multiple: Story<Props> = Template.bind({});
Multiple.args = {
    ...Basic.args,
    permission: ['granted', 'pageXYZ']
};

export const Wildcard: Story<Props> = Template.bind({});
Wildcard.args = {
    ...Basic.args,
    permission: 'pageXYZ'
};

export const Denied: Story<Props> = Template.bind({});
Denied.args = {
    ...Basic.args,
    permission: 'denied'
};

export const AnyDenied: Story<Props> = Template.bind({});
AnyDenied.args = {
    ...Basic.args,
    permission: ['denied', 'granted']
};
