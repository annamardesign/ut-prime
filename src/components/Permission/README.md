# Permission

Render children when permission is granted

## How to use

```jsx
import Permission from 'ut-front-devextreme/core/Permission';

<Permission permission='x.y.z'>content when x.y.z is allowed</Permission>
<Permission permission={['a.b.c', 'x.y.z']}>content when a.b.c and x.y.z are allowed</Permission>
```

## Props

- **permission** - permission(s) to check

| propName   | propType          | defaultValue | isRequired |
| ---------- | ----------------- | ------------ | ---------- |
| permission | string | string[] |              | no         |
