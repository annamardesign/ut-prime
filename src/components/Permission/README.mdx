# Permission
import { Canvas, Story, ArgsTable } from "@storybook/addon-docs";

Render children when permission is granted

## How to use

```jsx
import Permission from 'ut-prime/core/Permission';

<Permission permission='x.y.z'>content when x.y.z is allowed</Permission>
<Permission permission={['a.b.c', 'x.y.z']}>content when a.b.c and x.y.z are allowed</Permission>
```

<Canvas>
    <Story id="permission--basic" />
</Canvas>

## Properties

<ArgsTable story="."/>
