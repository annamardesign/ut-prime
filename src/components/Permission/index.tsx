import React from 'react';

import { StyledType } from './Permission.types';
import usePermission from '../hooks/usePermission';

const Permission: StyledType = ({ permission, children }) => usePermission(permission) ? <>{children}</> : null;

export default Permission;
