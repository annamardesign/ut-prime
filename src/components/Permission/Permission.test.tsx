import React from 'react';

import { render } from '../test';
import { Basic, Granted, Multiple, Wildcard, AnyDenied, Denied } from './Permission.stories';

describe('<Permission />', () => {
    it('Basic render equals snapshot', async() => {
        const { findByTestId } = render(<Basic {...Basic.args} />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
    it('Granted render equals snapshot', async() => {
        const { findByTestId } = render(<Granted {...Granted.args} />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
    it('Multiple render equals snapshot', async() => {
        const { findByTestId } = render(<Multiple {...Multiple.args} />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
    it('Wildcard render equals snapshot', async() => {
        const { findByTestId } = render(<Wildcard {...Wildcard.args} />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
    it('AnyDenied render equals snapshot', async() => {
        const { findByTestId } = render(<AnyDenied {...AnyDenied.args} />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
    it('Denied render equals snapshot', async() => {
        const { findByTestId } = render(<Denied {...Denied.args} />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
});
