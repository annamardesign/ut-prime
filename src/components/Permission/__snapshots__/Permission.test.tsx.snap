// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Permission /> AnyDenied render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
/>
`;

exports[`<Permission /> Basic render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  Permissions checked: 
  undefined
</div>
`;

exports[`<Permission /> Denied render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
/>
`;

exports[`<Permission /> Granted render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  Permissions checked: 
  granted
</div>
`;

exports[`<Permission /> Multiple render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  Permissions checked: 
  granted,pageXYZ
</div>
`;

exports[`<Permission /> Wildcard render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  Permissions checked: 
  pageXYZ
</div>
`;
